File: iOS/CMIICollector/Study.swift
//
//  Study.swift
//  Trial vocabulary, the play, and its seeded generator.
//
//  The play carries a FULLY MATERIALISED trial list rather than parameters to
//  be re-generated. Swift's and Kotlin's seeded PRNGs disagree, so generating per
//  platform would silently produce different designs from the same seed and break
//  cross-platform comparability (GESTURE_STUDY_SPEC.md §9b). Both apps play back
//  a stored list; SplitMix64 below is specified explicitly for the same reason.
//
import Foundation
import UIKit

/// Portable PRNG — identical output for a given seed on any platform.
struct SplitMix64: RandomNumberGenerator {
    private var state: UInt64
    init(seed: UInt64) { state = seed }
    mutating func next() -> UInt64 {
        state = state &+ 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }
}

/// Cardinal directions only.
///
/// Scenes no longer decode INTO this - a scene's `dir` is a plain string, because
/// the book also has IN/OUT and CW/CCW and a closed enum would fail to decode
/// them outright. This survives for the offline generator and for the geometry
/// check, which is only meaningful for the four cardinals anyway.
enum GDir: String, Codable, CaseIterable {
    case L, R, U, D
    var arrow: String {
        switch self {
        case .L: return "arrow.left"
        case .R: return "arrow.right"
        case .U: return "arrow.up"
        case .D: return "arrow.down"
        }
    }
    var word: String {
        switch self {
        case .L: return "left"
        case .R: return "right"
        case .U: return "up"
        case .D: return "down"
        }
    }
}

/// The built-in vocabulary.
///
/// This is NO LONGER what a scene is typed as - scenes carry a gesture slug as a
/// plain string plus the verb and accepted labels the server sends, so a gesture
/// added to the book on the web appears on the tablet with no app release. The
/// enum survives for the offline fallback generator and as a label of last resort
/// for a payload that predates the book.
enum GType: String, Codable, CaseIterable {
    case tap, double_tap, long_press, swipe, drag

    /// Direction is part of the class: swipe-left and swipe-up are different
    /// movements at the wrist, so they are scheduled and balanced separately.
    var directions: [GDir] { (self == .swipe || self == .drag) ? GDir.allCases : [] }

    var verb: String {
        switch self {
        case .tap:        return "Tap"
        case .double_tap: return "Tap twice"
        case .long_press: return "Press and hold"
        case .swipe:      return "Flick"
        case .drag:       return "Slowly drag"
        }
    }

    /// Gesture labels the on-device classifier may legitimately produce. Used only
    /// to VERIFY the trial — the cue is the label (see spec §2).
    var acceptedLabels: [String] {
        switch self {
        case .tap:        return ["tap"]
        case .double_tap: return ["tap"]          // two of them; handled separately
        case .long_press: return ["long_press"]
        case .swipe:      return ["swipe"]
        // Only "scroll" (slow displacement). Accepting "swipe" here made the match
        // flag meaningless: session 0824_0921 trial 0 cued a slow drag, the participant
        // flicked (200 ms / 167 px vs 1025-1117 ms for the real drags), and it still
        // scored as a match. The separation is clean, so the check should be strict.
        case .drag:       return ["scroll"]
        }
    }
}

/// One scene: a gesture cue, the grid it is shown in, the key block the
/// participant must act on, and how long they get.
///
/// rows/cols/durationMs are optional so a play generated before scenes
/// existed - including one sitting in the on-disk cache - still decodes; the
/// play-level values fill in.
/// One entry in a web scene's menu.
struct WebSite: Codable, Hashable, Identifiable {
    var label: String
    var url: String
    var id: String { url }
    /// Nil for anything that is not plain https - the tile is dropped rather than
    /// offering the participant a page that will not load.
    var link: URL? {
        guard let u = URL(string: url), u.scheme == "https" else { return nil }
        return u
    }
}

struct Trial: Codable, Identifiable {
    var i: Int
    /// Gesture slug from the book, e.g. "tap" or something added later.
    var type: String
    /// Direction code: L/R/U/D, or IN/OUT, CW/CCW for gestures that use them.
    var dir: String?
    var row: Int
    var col: Int
    var picture: String
    var rows: Int? = nil
    var cols: Int? = nil
    var durationMs: Int? = nil
    /// Sent by the server so an unknown gesture still renders and still scores.
    var verb: String? = nil
    var directional: Bool? = nil
    var acceptedLabels: [String]? = nil
    var tag: String? = nil
    var dirWord: String? = nil
    var dirIcon: String? = nil
    /// False for IN/OUT and CW/CCW: those cannot be checked against the stroke
    /// geometry, so the label alone decides the match.
    var dirVerifiable: Bool? = nil
    var affordance: String? = nil
    /// A travelling gesture ends on a different block. Drag does; a flick is
    /// ballistic and a scroll stays inside its own block.
    /// An off-screen scene asks for something away from the tablet - reaching for
    /// the water bottle, resting. The board is shadowed and inert and the screen
    /// carries the prompt instead of a cue.
    var offscreen: Bool? = nil
    /// A web scene hands the whole screen to a pinned page for its slot. The
    /// participant uses a real site; touches still land in this app's window, so
    /// they are still recorded, and the app stays foreground so the BLE beacon
    /// keeps advertising - neither of which is true of a real third-party app.
    var web: Bool? = nil
    var url: String? = nil
    /// A web scene may offer a menu instead of a single page. Free choice is part
    /// of what makes the session naturalistic; a bounded list keeps participants
    /// comparable and keeps uncurated content off the screen.
    var sites: [WebSite]? = nil
    var prompt: String? = nil
    var travels: Bool? = nil
    var toRow: Int? = nil
    var toCol: Int? = nil
    var toPicture: String? = nil
    /// Fixed-slot timing from the server: where this scene sits on the play's own
    /// timeline, and how long its slot is. Present from playVersion 3.
    var startMs: Int? = nil
    var slotMs: Int? = nil
    var gapMs: Int? = nil
    var id: Int { i }

    private var builtIn: GType? { GType(rawValue: type) }
    /// The cardinal direction, when this scene has one that geometry can check.
    var cardinal: GDir? { dir.flatMap { GDir(rawValue: $0) } }

    var directionWord: String {
        if let w = dirWord, !w.isEmpty { return w }
        return cardinal?.word ?? (dir ?? "").lowercased()
    }
    /// Validated on device: the server picks the symbol but cannot know what this
    /// iOS version renders, and an unknown name draws blank space next to the cue.
    var directionIcon: String? {
        if let i = dirIcon, !i.isEmpty, UIImage(systemName: i) != nil { return i }
        if let a = cardinal?.arrow, UIImage(systemName: a) != nil { return a }
        return nil
    }

    /// What the tablet shows. Falls back to the built-in verb, then the raw slug
    /// with underscores opened up, so a new gesture is never a blank cue.
    var displayVerb: String {
        if let v = verb, !v.isEmpty { return v }
        if let b = builtIn { return b.verb }
        return type.replacingOccurrences(of: "_", with: " ").capitalized
    }

    /// Classifier outputs that count as performing this gesture. Empty means the
    /// gesture cannot be auto-verified, and match is recorded as unknown rather
    /// than scoring every trial a failure.
    var labels: [String] {
        if let l = acceptedLabels { return l }
        return builtIn?.acceptedLabels ?? []
    }

    func grid(default s: Play) -> (rows: Int, cols: Int) {
        (max(1, rows ?? s.rows), max(1, cols ?? s.cols))
    }
    func block(default s: Play) -> Int { row * grid(default: s).cols + col }
    var isOffscreen: Bool { offscreen == true }
    var isWeb: Bool { web == true && (webURL != nil || !siteList.isEmpty) }
    var siteList: [WebSite] { (sites ?? []).filter { $0.link != nil } }
    var webURL: URL? {
        guard let u = url, !u.isEmpty, let parsed = URL(string: u),
              parsed.scheme == "https" else { return nil }
        return parsed
    }
    /// Neither kind of scene cues a gesture on the board, so both run their slot
    /// out on the clock instead of ending on a response.
    var isFreeform: Bool { isOffscreen || isWeb }
    var promptText: String {
        let p = (prompt ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return p.isEmpty ? displayVerb : p
    }
    var isTravelling: Bool { travels == true && toRow != nil && toCol != nil }

    var cueText: String {
        if isOffscreen { return promptText }
        // A travelling scene shows its destination on screen, so naming the
        // direction as well is redundant - the target says where to go.
        if isTravelling { return displayVerb }
        guard dir != nil else { return displayVerb }
        return "\(displayVerb) \(directionWord)"
    }
}

enum Preset: String, CaseIterable, Identifiable {
    case demo, standard, full
    var id: String { rawValue }

    /// How many distinct blocks each (gesture, direction) condition is shown in.
    var blocksPerCondition: Int {
        switch self {
        case .demo: return 1
        case .standard: return 2
        case .full: return 4
        }
    }
    var title: String {
        switch self {
        case .demo:     return "Demo · 11"
        case .standard: return "Standard · 22"
        case .full:     return "Full · 44"
        }
    }
}

enum Animals {
    /// Filtered at runtime so a symbol missing on this iOS version cannot render blank.
    static let all: [String] = {
        let candidates = ["hare.fill", "tortoise.fill", "bird.fill", "fish.fill",
                          "ladybug.fill", "ant.fill", "lizard.fill", "cat.fill",
                          "dog.fill", "teddybear.fill", "pawprint.fill"]
        let ok = candidates.filter { UIImage(systemName: $0) != nil }
        return ok.count >= 4 ? ok : ["pawprint.fill", "hare.fill", "tortoise.fill", "ant.fill"]
    }()
}

struct Play: Codable {
    var playVersion = 3
    var name: String
    var seed: UInt64
    var rows = 2
    var cols = 2
    var readyMs = 800
    var cueTimeoutMs = 6000
    var gapMinMs = 1200
    var gapMaxMs = 2500
    var settleMs = 450
    var layouts: [String]? = nil
    var blockPictures: [String]
    var trials: [Trial]

    var count: Int { trials.count }

    /// Balanced: each (gesture, direction) condition appears in a different block on
    /// each repetition, so gesture type is never tied to screen position — at the
    /// wrist, reach location would otherwise dominate gesture identity (spec §2).
    static func make(preset: Preset, seed: UInt64) -> Play {
        var rng = SplitMix64(seed: seed)

        var pics = Animals.all
        pics.shuffle(using: &rng)
        let blockPics = Array(pics.prefix(4))

        var conditions: [(GType, GDir?)] = []
        for t in GType.allCases {
            if t.directions.isEmpty { conditions.append((t, nil)) }
            else { for d in t.directions { conditions.append((t, d)) } }
        }

        var out: [Trial] = []
        for rep in 0..<preset.blocksPerCondition {
            for (ci, c) in conditions.enumerated() {
                let b = (ci + rep) % 4
                out.append(Trial(i: 0, type: c.0.rawValue, dir: c.1?.rawValue,
                                 row: b / 2, col: b % 2, picture: blockPics[b],
                                 rows: 2, cols: 2,
                                 verb: c.0.verb,
                                 directional: !c.0.directions.isEmpty,
                                 acceptedLabels: c.0.acceptedLabels,
                                 tag: "normal",
                                 dirWord: c.1?.word, dirIcon: c.1?.arrow,
                                 dirVerifiable: true, affordance: "image"))
            }
        }
        out.shuffle(using: &rng)
        for i in out.indices { out[i].i = i }

        return Play(name: preset.rawValue, seed: seed,
                        blockPictures: blockPics, trials: out)
    }

    /// Swap out any SF Symbol this iOS version does not have.
    ///
    /// Block pictures are chosen on the server, which cannot know what this
    /// device can render - an unknown symbol name draws nothing, and the
    /// participant would be cued to tap an empty square.
    func sanitised() -> Play {
        var out = self
        let ok = Animals.all
        var swaps: [String: String] = [:]
        out.trials = trials   // replaced below if anything needs swapping
        out.blockPictures = blockPictures.enumerated().map { idx, nameIn in
            if UIImage(systemName: nameIn) != nil { return nameIn }
            let sub = ok[idx % ok.count]
            swaps[nameIn] = sub
            return sub
        }
        if !swaps.isEmpty {
            out.trials = trials.map { t in
                var t = t
                if let sub = swaps[t.picture] { t.picture = sub }
                if let tp = t.toPicture, let sub = swaps[tp] { t.toPicture = sub }
                return t
            }
        }
        return out
    }

    func jsonData() -> Data? {
        let e = JSONEncoder()
        e.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try? e.encode(self)
    }
}

File: Android/app/src/main/java/com/cmii/collector/Study.kt
package com.cmii.collector

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * The play, exactly as the server ships it.
 *
 * This mirrors the iOS Study.swift because the JSON is the contract between the
 * two platforms - the same play.json has to drive an iPad and this tablet, and
 * the CSVs they write have to be read by one analysis pipeline. Where the two
 * apps must agree, they agree here.
 *
 * Every field the server may omit is nullable with a default: an older payload,
 * or a scene type this build does not know, must not fail to parse. A tablet
 * that refuses to start because of an unexpected key is worse than one that
 * renders a scene plainly.
 */

@Serializable
data class WebSite(
    val label: String = "",
    val url: String = ""
) {
    /** Null for anything that is not plain https - the tile is dropped rather
     *  than offering the participant a page that will not load. */
    val link: String?
        get() = url.takeIf { it.startsWith("https://") }
}

@Serializable
data class Trial(
    val i: Int = 0,
    val type: String = "",
    val dir: String? = null,
    val device: Int = 0,
    val row: Int = 0,
    val col: Int = 0,
    val rows: Int? = null,
    val cols: Int? = null,
    val picture: String? = null,
    val durationMs: Int? = null,
    val slotMs: Int? = null,
    val startMs: Int? = null,
    val gapMs: Int? = null,
    val verb: String? = null,
    val directional: Boolean? = null,
    val affordance: String? = null,
    val dirWord: String? = null,
    val dirIcon: String? = null,
    val dirVerifiable: Boolean? = null,
    val acceptedLabels: List<String> = emptyList(),
    val tag: String? = null,
    val travels: Boolean? = null,
    val toRow: Int? = null,
    val toCol: Int? = null,
    val toPicture: String? = null,
    val prompt: String? = null,
    /** Away from the tablet entirely - reaching for the water bottle. */
    val offscreen: Boolean? = null,
    /** A slot this tablet does not own; the other one is playing. */
    val empty: Boolean? = null,
    /** A pinned page, or a menu of them. */
    val web: Boolean? = null,
    val url: String? = null,
    val sites: List<WebSite> = emptyList()
) {
    val isOffscreen: Boolean get() = offscreen == true
    val isWaiting: Boolean get() = empty == true
    val isWeb: Boolean get() = web == true && (webUrl != null || siteList.isNotEmpty())

    val webUrl: String? get() = url?.takeIf { it.startsWith("https://") }
    val siteList: List<WebSite> get() = sites.filter { it.link != null }

    /**
     * Neither a waiting slot, an off-screen prompt nor a web page cues a gesture
     * on the board, so all three run their slot out on the clock rather than
     * ending when someone happens to touch the screen.
     */
    val isFreeform: Boolean get() = isOffscreen || isWaiting || isWeb

    /**
     * Whether this scene should show the card before the gesture.
     *
     * A drag and a flick both act on a particular card and move it somewhere -
     * out of the deck, or off the board - so the participant has to see which
     * card that is. A tap does not: the card has to start face down or there is
     * nothing to flip, and the flip is the gesture being measured.
     */
    val revealsCard: Boolean get() = isTravelling || type == "swipe"

    val promptText: String get() = prompt.orEmpty()

    val displayVerb: String
        get() = (verb ?: type).replaceFirstChar { it.uppercase() }

    val directionWord: String get() = dirWord ?: dir.orEmpty().lowercase()

    val isTravelling: Boolean get() = travels == true && toRow != null && toCol != null

    fun grid(default: Play): Pair<Int, Int> =
        Pair(rows ?: default.rows, cols ?: default.cols)

    /** How long this scene occupies, falling back the way the server does. */
    fun slot(play: Play): Int =
        slotMs ?: (play.readyMs + (durationMs ?: play.cueTimeoutMs) + play.settleMs)
}

@Serializable
data class Play(
    val playVersion: Int = 3,
    val name: String = "",
    val seed: Long = 0,
    val rows: Int = 2,
    val cols: Int = 2,
    val readyMs: Int = 800,
    val cueTimeoutMs: Int = 6000,
    val gapMinMs: Int = 1200,
    val gapMaxMs: Int = 2500,
    val settleMs: Int = 450,
    val layouts: List<String>? = null,
    val blockPictures: List<String> = emptyList(),
    val trials: List<Trial> = emptyList(),
    val totalMs: Int = 0,
    val waitingText: String? = null,
    /** Present only on a half handed to one tablet. */
    val device: Int? = null,
    val tabletLabel: String? = null,
    val experimentId: Int? = null,
    val experimentName: String? = null,
    val scheduledStartMs: Long? = null,
    val playId: Int? = null
) {
    val count: Int get() = trials.size
    val waiting: String get() = waitingText?.takeIf { it.isNotBlank() } ?: "Please read the other Device"
}

@Serializable
data class ExperimentInfo(
    val id: Int,
    val name: String = "",
    val description: String = "",
    val sessions: Int = 0,
    val playId: Int? = null,
    val playName: String? = null,
    val trialCount: Int = 0,
    val totalMs: Int = 0,
    val tablets: Int = 1,
    val notes: String? = null,
    val startISO: String? = null,
    // The sitting, described once on the server so two tablets cannot disagree.
    val participant: String = "",
    val watchWrist: String = "",
    val interactingHand: String = "",
    val posture: String = "",
    val tabletOrientation: String = "",
    val missing: List<String> = emptyList(),
    // Which device plays which half. A tablet matches its own id against these
    // rather than being told its role by hand.
    val tabletA: String? = null,
    val tabletB: String? = null,
    val tabletALabel: String? = null,
    val tabletBLabel: String? = null,
    val advertiseA: String? = null,
    val advertiseB: String? = null,
    /** "B", "A" or "both" — which tablet the watch scans for in this experiment. */
    val beacon: String = "B",
    /**
     * A testing experiment: every tablet lists it, it needs no schedule and no
     * assignment, and its sessions are not participant data. Shown as such on
     * the list, because it is the one experiment that skips every check.
     */
    val free: Boolean = false
) {
    /**
     * Whether THIS tablet is a beacon here.
     *
     * A single-tablet run always advertises: there is no decoy to contrast
     * against, and a silent lone tablet leaves the watch with nothing.
     */
    fun advertisesFor(deviceId: String): Boolean {
        if (tablets <= 1) return true
        // Free and unassigned: nobody is the decoy, so whoever picked it up is
        // the beacon. Otherwise a test looks like a BLE fault.
        if (free && tabletA == null && tabletB == null) return true
        val mine = roleFor(deviceId) ?: return false
        return beacon == "both" || beacon == mine
    }

    val hasAssignment: Boolean get() = tabletA != null || tabletB != null

    /** This device's half, or null when it is not one of the assigned tablets. */
    fun roleFor(deviceId: String): String? = when (deviceId) {
        tabletA -> "A"
        tabletB -> "B"
        else -> null
    }

    /** The advertise name for this device's half, as the server holds it. */
    fun advertiseFor(deviceId: String): String? = when (deviceId) {
        tabletA -> advertiseA
        tabletB -> advertiseB
        else -> null
    }

    /**
     * What this tablet is for THIS experiment, or null if it cannot run it.
     *
     * Assigned: the server decides. Unassigned single-tablet: there is no decoy,
     * so the one tablet is the beacon. Unassigned two-tablet: refused - neither
     * tablet can know which half it is, and guessing would either silence the
     * beacon or duplicate it.
     */
    fun resolvedRole(deviceId: String): String? {
        // A free experiment is offered to every tablet, assigned or not. An
        // assignment is still honoured where one exists, so a free two-tablet
        // play can still be split deliberately.
        if (free) return roleFor(deviceId) ?: if (tablets <= 1) "B" else "A"
        // An experiment with no device assigned is not ready, whatever else is
        // set on it, so no tablet offers it.
        val mine = roleFor(deviceId) ?: return null
        // A and B only mean anything when there are two. With one tablet there is
        // no decoy to contrast against, so it is the beacon whichever slot it was
        // assigned to - otherwise a single-tablet run assigned to slot A would go
        // silent and the watch would record nothing at all.
        return if (tablets <= 1) "B" else mine
    }

    val assignedTo: String
        get() = listOfNotNull(
            tabletALabel?.let { "A: $it" }, tabletBLabel?.let { "B: $it" }
        ).joinToString(" · ")

    val hasPlay: Boolean get() = (playId ?: 0) > 0 && trialCount > 0

    /** Why Start is unavailable, in the operator's terms rather than the schema's. */
    val blockedReason: String
        get() = when {
            (playId ?: 0) == 0 -> "no play assigned"
            trialCount == 0 -> "its play has no scenes"
            else -> ""
        }

    val durationText: String
        get() {
            val s = totalMs / 1000
            return if (s >= 60) "${s / 60} min" + if (s % 60 == 0) "" else " ${s % 60} s"
            else "$s s"
        }

    /** Epoch millis of the scheduled start, or null. */
    val startAtMs: Long?
        get() = startISO?.takeIf { it.isNotBlank() }?.let { Iso.parse(it) }
}

@Serializable
data class ExperimentList(val experiments: List<ExperimentInfo> = emptyList())
